cmd/fs_handlers.go

```diff
@@ -77,6 +78,39 @@ func handleStageDirectory(ctx *cli.Context, ctl *client.Client, root, repoRoot s
 	err := filepath.Walk(root, func(childPath string, info os.FileInfo, err error) error {
 		repoPath := filepath.Join("/", repoRoot, childPath[len(root):])
 
+		if info.Mode() & os.ModeSymlink != 0 {
+			resolvedPath, err := filepath.EvalSymlinks(childPath)
+			if err != nil {
+				if os.IsNotExist(err) {
+					log.Warningf("%v symlink points to non existing file or directory", childPath)
+					return nil
+				}
+				if err.Error() == "EvalSymlinks: too many links" {
+					log.Warningf("%v too many links needed to be followed", childPath)
+					return nil
+				}
+				return fmt.Errorf("Failed to resolve: %v: %v", childPath, err)
+			}
+			info, err = os.Stat(resolvedPath)
+			if err != nil {
+				if os.IsNotExist(err) {
+					log.Warningf("%v resolved %v which points to non existing file or directory", childPath, resolvedPath)
+					return nil
+				}
+				return fmt.Errorf("Failed to do os.Stat(%v): %v", resolvedPath, err)
+			}
+			if !info.Mode().IsRegular() {
+				// We are staging only links which resolve in regular files
+				// Since Walk does not travel symlinks, there is no point
+				// to stage links pointing to a directory. None of its files
+				// will be walked and staged.
+				log.Warningf("Not staging %v which point to a non regular file: %v %v", childPath, info.Mode(), resolvedPath)
```

Comment:
Isn't that kinda inconsistent? We follow symlinks for files, but not directories? That doesn't really follow the principle of least surprise. Also, I would prefer to have an option to ignore symlinks altogether, which disables this behavior. Sometimes I don't want to follow symbolic links that might me take elsewhere.

Reply:
Yep, I am not proud of it. The problem is that  `filepath.Walk` by design does not follow symlinks to the directories. Fixing it would require a use of a different library.

Shall I dig in that direction?

I will work on ignore symlinks option as well.

Reply:
> Fixing it would require a use of a different library.

Why that? When you encounter a (valid) symbolic link to a directory, you simply call ``filepath.Walk`` on that resolved directory path. I agree that it might get a bit clumsy, but it will work. A better option might be ``ioutil.ReadDir()`` and use that in a recursive function: When encountering a directory (either regular one or resolved from a symbolic link) one would call the function that uses ``ioutil.ReadDir()``. 

Care must be taken to limit the recursion depth, since symlinks can result in loops. But that applies to filepath.Walk too...

Reply:
redone in series of patches. It is not as elegant as I wished. But it also simplified staging logic: no need to create dirs, internal `Stage` creates all necessary parent folders. Also all traveling is in one place.

Side effect: empty dirs will not be created. Do we want them? It would be relatively quick fix.